src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs

```diff
@@ -46,11 +77,17 @@ public AuthorizeFilter(AuthorizationPolicy policy)
                 throw new ArgumentNullException(nameof(context));
             }
 
+            var effectivePolicy = Policy ?? await AuthorizationPolicy.CombineAsync(PolicyProvider, AuthorizeData);
```

Comment:
I have the same concern as @pranavkm. You can technically implement your own IPolicyProvider and get a reference to the IServiceProvider in the constructor (which is ugly, but possible) and get access to the request for example there, using IHttpContextAccessor or something similar.

What I'm most concerned about is this double behavior of cache the policy if you use one constructor and do not if you use the other constructor. We can remove caching all together, and it will be cleaner, but I would ask what would be the cost of computing the policy per request given the most common case in which the policy is "static" (always the same, based on Options). Should the policy contain information about it's nature? (or the result from CombineAsync) that indicates whether or not the policy can be cached.